apps/builder/app/builder/features/style-panel/sections/backgrounds/background-layers.ts

```diff
@@ -0,0 +1,252 @@
+import type {
+  CreateBatchUpdate,
+  DeleteProperty,
+  StyleUpdateOptions,
+} from "../../shared/use-style-data";
+
+export const layeredBackgroundPropsDefaults = {
+  backgroundAttachment: { type: "keyword", value: "scroll" },
+  backgroundClip: { type: "keyword", value: "border-box" },
+  backgroundBlendMode: { type: "keyword", value: "normal" },
+  backgroundImage: { type: "keyword", value: "none" },
+  backgroundOrigin: { type: "keyword", value: "padding-box" },
+  backgroundPositionX: { type: "unit", value: 0, unit: "%" },
+  backgroundPositionY: { type: "unit", value: 0, unit: "%" },
+  backgroundRepeat: { type: "keyword", value: "repeat" },
+  backgroundSize: { type: "keyword", value: "auto" },
+} as const satisfies Record<string, BackgroundStyleValue>;
+
+export type BackgroundStyleValue = LayersValue["value"][number];
+
+export const isBackgroundStyleValue = (
+  style: StyleValue
+): style is BackgroundStyleValue => {
+  if (
+    style.type === "unit" ||
+    style.type === "keyword" ||
+    style.type === "unparsed" ||
+    style.type === "image" ||
+    style.type === "invalid"
+  ) {
+    return true;
+  }
+  return false;
+};
+
+export const layeredBackgroundProps = Object.keys(
+  layeredBackgroundPropsDefaults
+) as (keyof typeof layeredBackgroundPropsDefaults)[];
+
+export type LayeredBackgroundProperty = (typeof layeredBackgroundProps)[number];
+
+export const isBackgroundLayeredProperty = (
+  prop: string
+): prop is LayeredBackgroundProperty => {
+  return layeredBackgroundProps.includes(prop as LayeredBackgroundProperty);
+};
+
+const getPropertyLayerCount = (style: StyleValueInfo) => {
+  if (style.value.type === "layers") {
+    return style.value.value.length;
+  }
+  return 0;
+};
+
+export const getLayerCount = (style: StyleInfo) => {
+  return Math.max(
+    ...layeredBackgroundProps.map((prop) => {
+      const styleValue = style[prop];
+
+      if (styleValue !== undefined) {
+        return getPropertyLayerCount(styleValue);
+      }
+
+      return 0;
+    })
+  );
+};
+
+export const getLayerBackgroundStyleInfo = (
+  layerNum: number,
+  style: StyleInfo
+): StyleInfo => {
+  const layerCount = getLayerCount(style);
+  if (layerNum >= layerCount) {
+    throw new Error(`${layerNum} is out of bounds`);
+  }
+
+  const result: StyleInfo = {};
+
+  for (const [property, value] of Object.entries(
+    layeredBackgroundPropsDefaults
+  )) {
+    result[property as LayeredBackgroundProperty] = { value };
+  }
+
+  for (const property of layeredBackgroundProps) {
+    const styleValue = style[property];
+
+    const valueStyle = styleValue?.value;
+    const localStyle = styleValue?.local;
+    const cascadedStyle = styleValue?.cascaded;
+
+    if (valueStyle?.type === "layers") {
+      const styleValue = valueStyle.value[layerNum];
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      result[property]!["value"] = styleValue;
```

Comment:
you could have avoided it, I bet it doesn't know what type you are operating on, so it can't tell if there is a value property

Reply:
all of these non-null assertions look fishy to me, I bet you can do better than this